app/lib/provider_discovery.rb

```diff
@@ -0,0 +1,38 @@
+# frozen_string_literal: true
+
+class ProviderDiscovery < OEmbed::ProviderDiscovery
+  USER_AGENT = "#{HTTP::Request::USER_AGENT} (Mastodon/#{Mastodon::VERSION}; +http://#{Rails.configuration.x.local_domain}/)"
+
+  class << self
+    def discover_provider(url, options = {})
+      res    = http_client.get(url)
+      format = options[:format]
+
+      raise OEmbed::NotFound, url if res.code != 200 || res.mime_type != 'text/html'
+
+      html = Nokogiri::HTML(res.to_s)
+
+      if format.nil? || format == :json
+        provider_endpoint = html.at_xpath('//link[@type="application/json+oembed"]')['href']
+        format ||= :json
+      end
+
+      if format.nil? || format == :xml
```

Comment:
I'm not sure this series of conditionals is doing what you expect it to do. Are you trying to do something like this? currently it uses JSON, or XML, or JSON if no format is specified.

```ruby
def provider_endpoint(format)
  case format
  when :json
    html.at_xpath('//link[@type="application/json+oembed"]')['href']
  when :xml
    html.at_xpath('//link[@type="application/json+oembed"]')['href']
  else
    return provider_endpoint(:json) || provider_endpoint(:xml)
  end
end
```